### ISSUE TITLE
Shutdown logic error in MamaQueueTestC.DispatchManyQueuesWithDispatchers

### ISSUE THREAD

**Issue by dpauls** [OpenMAMA/OpenMAMA#216]
# ISSUE_TITLE
## Summary
I should mention this problem was observed via code inspection of the unit test code for MamaQueueTestC.DispatchManyQueuesWithDispatchers while investigating the problem reported in https://github.com/OpenMAMA/OpenMAMA/issues/215.  So I have not observed this problem.  But based on my interpretation of the code, I believe it could be a problem.

It seems to me that the intent of the test is to wait for all events to have been dispatched from all queues, and then destroy the dispatchers and queues.  Instead, it appears the test waits for one queue to dispatch all of its events, and then the test destroys all dispatchers and queues.

I wonder if it is possible for a queue to have events remaining on it when it is destroyed, which could lead to an error trying to destroy the queue?

## Areas Affected
- [ ] MAMAC
- [ ] MAMACPP
- [ ] MAMADOTNET
- [ ] MAMAJNI
- [ ] MAMDA
- [ ] MAMDACPP
- [ ] MAMDADOTNET
- [ ] MAMDAJNI
- [ ] Visual Studio
- [ ] SCons
- [X] Unit Tests
- [ ] Examples

## Details
The MamaQueueTestC.DispatchManyQueuesWithDispatchers test creates 10 queues and dispatches 100 events to each queue.  In OpenMAMA 2.4.1, the tests waited for 1000 events to be processed off of the queues in aggregate.  In OpenMAMA 6.1.0-rc1, the test is a bit more precise in that it records the number of events dispatched from each queue separately.  Prior to destroying dispatchers and queues, I would have expected the test to wait for all events to be dispatched from all queues.  However, it appears as though it begins the shutdown procedure after a single queue has dispatched all of its events.

I believe the following patch would address the issue.

```
Index: mama/c_cpp/src/gunittest/c/queuetest.cpp
===================================================================
--- mama/c_cpp/src/gunittest/c/queuetest.cpp    (revision 2940)
+++ mama/c_cpp/src/gunittest/c/queuetest.cpp    (working copy)
@@ -115,13 +115,19 @@
     mamaQueue_getClosure (queue, &pIndex);
     size_t index = (size_t)pIndex;

-    if (fixture->m_numEvents == ++fixture->m_numDispatches[index])
+    ++fixture->m_numDispatches[index];
+    int x;
+    for (x=0; x!=10; x++)
     {
-        for (int x=0; x!=10; x++)
+        if (fixture->m_numEvents != fixture->m_numDispatches[index]) break;
+    }
+
+    if (x==10) {
+        for (x=0; x!=10; x++)
         {
             mamaQueue_stopDispatch (fixture->m_queues[x]);
         }
-    wsem_post (&fixture->m_sem);
+        wsem_post (&fixture->m_sem);
     }
}
```

Again, it is worth noting I have not observed this problem -- this was picked up by code inspection only.  So it is possible I've misinterpreted the intent of the test or other details of what would happen when the test is running.

**Comment by fquinner** [comment]
`wsem_post` will increment the semaphore which is waited for in the unit test function and is called 10 times when each queue / dispatcher pair reaches its target. This is also called 10 times in the calling unit test function. so it should effectively block at this point until all 10 queues have posted:

```
    for (int x = 0; x!=m_numQueues; x++)
    {
        ASSERT_EQ(0, wsem_wait (&m_sem));
    }
```

Before this change, the test assumed that all queue events had fired before calling `mamaDispatcher_destroy` in the unit test (mamaDispatcher_destroy also calls the queue's stopDispatch function) before all events had come in. This could have led to false positives in the unit test as all events wouldn't necessarily have had to fire before exiting. I am not a fan of these sorts of tests because a queue which lost an event somewhere would end up deadlocking rather than returning an error and moving on, but that's another story.

**Comment by dpauls** [comment]
Frank, I agree with your assessment related to `wsem_wait` and `wsem_post`.  I missed the important detail that the test waits 10 times.

However, I still have a concern that all queues are told to stop dispatching as soon as one queue has received all events.  In onBgEvent():
```
    if (fixture->m_numEvents == ++fixture->m_numDispatches[index])
    {
        for (int x=0; x!=10; x++)
        {
            mamaQueue_stopDispatch (fixture->m_queues[x]);
        }
```

Shouldn't this just be:
```
    if (fixture->m_numEvents == ++fixture->m_numDispatches[index])
    {
        mamaQueue_stopDispatch (fixture->m_queues[index]);
```

It seems that by stopping queues that haven't finished dispatching events, some queues may never finish dispatching all of their events?  But as I've outlined in https://github.com/OpenMAMA/OpenMAMA/issues/215 however, `mamaQueue_stopDispatch()` without stopping the dispatcher doesn't really have an effect.

Given the issue with `mamaQueue_stopDispatch()`, this is probably a fairly low priority thing to deal with.  But it seems like a good change to make to the test anyways, in case `mamaQueue_stopDispatch()` is fixed in the future.

What do you think?

**Comment by fquinner** [comment]
Yeah agreed - should make both changes at once. Would also like to change that wsem_wait to a wsem_timedwait followed by some count checks though I know these tests are a deadlock minefield anyway.

**Comment by fquinner** [comment]
Fix pushed into next: 02fa7e45153152b62ebf603917fe25855ddc3e2d
